File: powersoft_ostec_customizations/doctype_changes/payment_entry_custom.py
import frappe
from frappe.utils import flt


def onload(self, method=None):
	"""Copy multi-year license renewal data when the Payment Entry is loaded (newly mapped document)."""
	if self.is_new():
		copy_multi_year_data_from_sales_invoice(self)


def before_insert(self, method=None):
	"""before_save is the reliable hook; before_insert is kept for completeness but does not run copy."""
	pass


def before_save(self, method=None):
	"""
	Copy multi-year data on first save of a new Payment Entry.
	Guarded so it only runs when the doc is new and the totals are not yet populated,
	preventing redundant re-copy on subsequent saves.
	"""
	if self.is_new() and not flt(self.custom_grand_total_year_2) and not flt(self.custom_grand_total_year_3):
		copy_multi_year_data_from_sales_invoice(self)


def copy_multi_year_data_from_sales_invoice(self):
	"""
	When a Payment Entry is created for a License Renewal Sales Invoice,
	carry forward all multi-year data:
	  - custom_quotation_type
	  - Year 2 and Year 3 totals, taxes, and grand totals
	"""
	try:
		sales_invoice_name = None
		for ref in self.get("references", []):
			if ref.reference_doctype == "Sales Invoice" and ref.reference_name:
				sales_invoice_name = ref.reference_name
				break

		if not sales_invoice_name:
			return

		try:
			sales_invoice = frappe.get_doc("Sales Invoice", sales_invoice_name)
		except frappe.DoesNotExistError:
			frappe.log_error(f"Payment Entry copy: Sales Invoice {sales_invoice_name} not found", "PE Multi-year copy")
			return

		# Only carry forward if this is a License Renewal
		if getattr(sales_invoice, "custom_quotation_type", None) != "License Renewal":
			return

		# --- Header Fields ---
		self.custom_quotation_type = sales_invoice.custom_quotation_type
		self.custom_total_year_2 = flt(sales_invoice.custom_total_year_2)
		self.custom_total_year_3 = flt(sales_invoice.custom_total_year_3)
		self.custom_total_taxes_and_charges_year_2 = flt(sales_invoice.custom_total_taxes_and_charges_year_2)
		self.custom_total_taxes_and_charges_year_3 = flt(sales_invoice.custom_total_taxes_and_charges_year_3)
		self.custom_grand_total_year_2 = flt(sales_invoice.custom_grand_total_year_2)
		self.custom_grand_total_year_3 = flt(sales_invoice.custom_grand_total_year_3)

		# --- Year 2 taxes child table ---
		self.set("custom_sales_taxes_and_charges_year_2", [])
		for tax in sales_invoice.custom_sales_taxes_and_charges_year_2:
			self.append("custom_sales_taxes_and_charges_year_2", {
				"charge_type": tax.charge_type,
				"row_id": tax.row_id,
				"description": tax.description,
				"account_head": tax.account_head,
				"rate": flt(tax.rate),
				"tax_amount": flt(tax.tax_amount),
				"total": flt(tax.total),
			})

		# --- Year 3 taxes child table ---
		self.set("custom_sales_taxes_and_charges_year_3", [])
		for tax in sales_invoice.custom_sales_taxes_and_charges_year_3:
			self.append("custom_sales_taxes_and_charges_year_3", {
				"charge_type": tax.charge_type,
				"row_id": tax.row_id,
				"description": tax.description,
				"account_head": tax.account_head,
				"rate": flt(tax.rate),
				"tax_amount": flt(tax.tax_amount),
				"total": flt(tax.total),
			})
	except Exception:
		frappe.log_error(frappe.get_traceback(), "Payment Entry: multi-year copy from Sales Invoice failed")

File: powersoft_ostec_customizations/public/js/payment_entry_custom.js
frappe.ui.form.on("Payment Entry", {
	onload(frm) {
		if (frm.is_new()) {
			copy_multi_year_data_from_invoice(frm);
		}
	},
	refresh(frm) {
		if (frm.is_new()) {
			copy_multi_year_data_from_invoice(frm);
		}
	}
});

function copy_multi_year_data_from_invoice(frm) {
	// Find referenced Sales Invoice
	let sales_invoice = null;
	if (frm.doc.references) {
		for (let ref of frm.doc.references) {
			if (ref.reference_doctype === "Sales Invoice" && ref.reference_name) {
				sales_invoice = ref.reference_name;
				break;
			}
		}
	}

	if (!sales_invoice) return;

	// Fetch Sales Invoice data and copy it
	frappe.db.get_doc("Sales Invoice", sales_invoice).then(invoice => {
		if (invoice.custom_quotation_type === "License Renewal") {
			frm.set_value("custom_quotation_type", invoice.custom_quotation_type);
			frm.set_value("custom_total_year_2", invoice.custom_total_year_2);
			frm.set_value("custom_total_year_3", invoice.custom_total_year_3);
			frm.set_value("custom_total_taxes_and_charges_year_2", invoice.custom_total_taxes_and_charges_year_2);
			frm.set_value("custom_total_taxes_and_charges_year_3", invoice.custom_total_taxes_and_charges_year_3);
			frm.set_value("custom_grand_total_year_2", invoice.custom_grand_total_year_2);
			frm.set_value("custom_grand_total_year_3", invoice.custom_grand_total_year_3);

			// Calculate sum of Year 1 (grand_total), Year 2, and Year 3 grand totals
			let total_all_years = flt(invoice.grand_total) + flt(invoice.custom_grand_total_year_2) + flt(invoice.custom_grand_total_year_3);
			
			frm.set_value("paid_amount", total_all_years);
			frm.set_value("received_amount", total_all_years);

			// Also update the allocated amount on the invoice reference line
			if (frm.doc.references) {
				for (let ref of frm.doc.references) {
					if (ref.reference_doctype === "Sales Invoice" && ref.reference_name === sales_invoice) {
						frappe.model.set_value(ref.doctype, ref.name, "allocated_amount", total_all_years);
						break;
					}
				}
			}

			// Clear and copy Year 2 taxes
			frm.clear_table("custom_sales_taxes_and_charges_year_2");
			if (invoice.custom_sales_taxes_and_charges_year_2) {
				for (let tax of invoice.custom_sales_taxes_and_charges_year_2) {
					let row = frm.add_child("custom_sales_taxes_and_charges_year_2");
					row.charge_type = tax.charge_type;
					row.row_id = tax.row_id;
					row.description = tax.description;
					row.account_head = tax.account_head;
					row.rate = tax.rate;
					row.tax_amount = tax.tax_amount;
					row.total = tax.total;
				}
			}

			// Clear and copy Year 3 taxes
			frm.clear_table("custom_sales_taxes_and_charges_year_3");
			if (invoice.custom_sales_taxes_and_charges_year_3) {
				for (let tax of invoice.custom_sales_taxes_and_charges_year_3) {
					let row = frm.add_child("custom_sales_taxes_and_charges_year_3");
					row.charge_type = tax.charge_type;
					row.row_id = tax.row_id;
					row.description = tax.description;
					row.account_head = tax.account_head;
					row.rate = tax.rate;
					row.tax_amount = tax.tax_amount;
					row.total = tax.total;
				}
			}

			frm.refresh_fields([
				"custom_quotation_type",
				"custom_total_year_2",
				"custom_total_year_3",
				"custom_total_taxes_and_charges_year_2",
				"custom_total_taxes_and_charges_year_3",
				"custom_grand_total_year_2",
				"custom_grand_total_year_3",
				"custom_sales_taxes_and_charges_year_2",
				"custom_sales_taxes_and_charges_year_3"
			]);
		}
	});
}
